taichi/runtime/llvm/runtime.cpp

```diff
@@ -642,19 +657,20 @@ void taichi_assert(Context *context, i32 test, const char *msg) {
   taichi_assert_runtime(context->runtime, test, msg);
 }
 
-const std::size_t ASSERT_MSG_BUFFER_SIZE = 2048;
-char assert_msg_buffer[ASSERT_MSG_BUFFER_SIZE];
-i32 assert_msg_buffer_lock = 0;
 void taichi_assert_format(LLVMRuntime *runtime, i32 test, const char *format,
                           ...) {
-  if (!enable_assert || test != 0)
+  if (!enable_assert || test != 0 || error_message.error_code)
     return;
   std::va_list args;
   va_start(args, format);
-  locked_task(&assert_msg_buffer_lock, [&] {
-    runtime->host_vsnprintf(assert_msg_buffer, ASSERT_MSG_BUFFER_SIZE, format,
-                            args);
-    runtime->assert_failed(assert_msg_buffer);
+  locked_task(&error_message.lock, [&] {
+    if (!error_message.error_code) {
+      error_message.error_code = 1;
+      runtime->host_vsnprintf(error_message.error_message_buffer,
+                              error_message.max_message_length,
+                              format,
+                              args);
```

Comment:
Is there anything we need to do here to terminate the thread and tell the host `error_code` is not 0?

Reply:
No, it should be the host's job to check the error code.